finality-aleph/src/sync/message_limiter.rs

```diff
@@ -0,0 +1,188 @@
+use std::fmt::{Display, Formatter};
+
+use parity_scale_codec::Encode;
+
+use crate::sync::data::MAX_SYNC_MESSAGE_SIZE;
+
+const MSG_BYTES_LIMIT: usize = MAX_SYNC_MESSAGE_SIZE as usize;
+
+pub struct Limiter<'a, D: Encode, const LIMIT: usize> {
+    msg: &'a [D],
+    start_index: usize,
+    indexes: Vec<usize>,
+}
+
+pub type MsgLimiter<'a, D> = Limiter<'a, D, MSG_BYTES_LIMIT>;
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum Error {
+    ItemTooBig,
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Error::ItemTooBig => write!(f, "Single item takes more than the limit"),
+        }
+    }
+}
+
+impl<'a, D: Encode, const LIMIT: usize> Limiter<'a, D, LIMIT> {
```

Comment:
for me this is fine, we would need to remember aboutadditional byte (or 2?) needed for encoding the vec size and I probably wouldnt mutate the msg but overall I like this approach much more. @lesniak43 wdyt, bs was because we wanted to clever and have no assumption about encoding but it looks like this is much better option

Reply:
done